### PR TITLE
Detect if loadPromise element already errored.

### PR DESCRIPTION
**Pull request by gmajoulet**
`loadPromise`, used as a response for many components `layoutCallback`, returns a promise that resolves or rejects once the element loads or errors.

It tries to detect if the passed element is already loaded, or sets load/error event listeners that will resolve/reject the returned promise.
But if the element already errored, the `loadPromise` will stay pending indefinitely. This PR adds support for checking if the element already errored.

The issue with HTMLMediaElement is that it is supposed to get an `error` property if it failed to load. However, it will only get it if the error comes from its `src` attribute. It the `<video>` has no `src`, but some `<source>` children, the `error` will not be available.
To clarify:
`<video src="404.mp4"></video>` will have `video.error` set to some `MediaError`.
`<video><source src="404.mp4"></video>` will not have `video.error` nor `source.error` set, and if you missed the `error` event, you can't check synchronously that the video errored.

To solve this, we add a custom property storing a boolean on the element itself, to mark it as errored.

Stories specific: as the media pool rotates the `HTMLMediaElement`, it tries to keep the new custom property as it re-uses elements.

This is used for stories to (1) detect that a video failed to play and display some error message on the UI, and (2) track video errors as part of our project to track video performance metrics.

TLDR: fixes the fact that `loadPromise` returned promise stays pending indefinitely if the element already fired its error events.

### CODE REVIEW

File: extensions/amp-story/1.0/media-tasks.js [hunk 1]
```diff
@@ -395,6 +419,7 @@ export class SwapIntoDomTask extends MediaTask {
 
     copyCssClasses(this.placeholderEl_, mediaEl);
     copyAttributes(this.placeholderEl_, mediaEl);
+    rotateWhitelistedProperties(this.placeholderEl_, mediaEl);
```
Comment (jridgewell) [inline extensions/amp-story/1.0/media-tasks.js]: Why are you doing this? Is this guaranteed to point to the same media?

File: extensions/amp-video/0.1/amp-video.js [hunk 2]
```diff
@@ -325,9 +325,16 @@ class AmpVideo extends AMP.BaseElement {
...
-    const promise = this.loadPromise(this.video_).then(() => {
-      this.element.dispatchCustomEvent(VideoEvents.LOAD);
-    });
+    const promise = this.loadPromise(this.video_).then(
+      () => {
+        this.element.dispatchCustomEvent(VideoEvents.LOAD);
+      },
+      () => {
+        // Ensures the load failure property is set on the right video element,
+        // as they can be swapped by the Stories media pool.
```
Comment (jridgewell) [inline extensions/amp-video/0.1/amp-video.js]: Can you explain? I don't think amp-video should be responsible for ever setting `LOAD_FAILURE_PROPERTY`.

File: extensions/amp-story/1.0/media-tasks.js (hunk 1 above)
Comment (gmajoulet) [reply extensions/amp-story/1.0/media-tasks.js]: > Why are you doing this?

There are restrictions on the number of `<video>` you can have on iOS, after a certain number of elements, they just don't play.
To solve this for Stories, the "Media pool" creates a small number of `<video>` elements and rotates through them. It updates its sources from the original video, and swaps it in the DOM.
This line ensures that, as the origin video is swapped for one from the media pool, it keeps the error property.

> Is this guaranteed to point to the same media?

Yes

File: extensions/amp-video/0.1/amp-video.js (hunk 2 above)
Comment (gmajoulet) [reply extensions/amp-video/0.1/amp-video.js]: I might a bit paranoid here, but given what I explained about the Stories media pool swapping video element, I'm worried about the following race condition:
1. `amp-video.layoutCallback` calls `loadPromise` with the original video
2. Stories media pool swaps the video for its own
3. `loadPromise` fails and sets `LOAD_FAILURE_PROPERTY` on the original video, that is no longer in the DOM

This line ensures that there's a 4rth step, where if the video got swapped between step (1) and (3), the error still gets applied to the right video element.
I haven't been able to repro that during local testing but I think it could happen in production.

File: extensions/amp-story/1.0/media-tasks.js (hunk 1 above)
Comment (jridgewell) [reply extensions/amp-story/1.0/media-tasks.js]: If we change the src of a video element, does it try to redownload the video? If so, we've reset the load.

File: extensions/amp-story/1.0/media-tasks.js [hunk 3]
```diff
@@ -43,6 +44,13 @@ const PROTECTED_CSS_CLASS_NAMES = [
  */
 const PROTECTED_ATTRIBUTES = ['id', 'src', 'class', 'autoplay'];
 
+/**
+ * Property names that should be propagated/removed from an element when
+ * swapping it into/out of the DOM.
+ * @const {!Array<string>}
+ */
+const WHITELISTED_PROPERTIES = [LOAD_FAILURE_PROPERTY];
```
[the remark this answers is not in the document]
Comment (gmajoulet) [reply extensions/amp-story/1.0/media-tasks.js]: I'm fine with any name, but `_REMOVE` might not be accurate as we still add it to the `toEl`. `PROPERTIES_TO_ROTATE` maybe?

File: extensions/amp-video/0.1/amp-video.js (hunk 2 above)
Comment (jridgewell) [reply extensions/amp-video/0.1/amp-video.js]: Ok. I don't think this is amp-video's responsibility then. It seems loadPromise should check that the media is still pointing at the same content before storing the error/success.

File: extensions/amp-story/1.0/media-tasks.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (newmuis) [reply extensions/amp-story/1.0/media-tasks.js]: `_ROTATE` works for me; we could also use `_COPY` as that's the language used further down where this actually happens

File: extensions/amp-story/1.0/media-tasks.js (hunk 1 above)
Comment (gmajoulet) [reply extensions/amp-story/1.0/media-tasks.js]: Good point. Cf comment on the main PR thread, I changed a few things

File: extensions/amp-video/0.1/amp-video.js (hunk 2 above)
Comment (gmajoulet) [reply extensions/amp-video/0.1/amp-video.js]: Got it. Removed it

File: extensions/amp-story/1.0/media-tasks.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (gmajoulet) [reply extensions/amp-story/1.0/media-tasks.js]: I changed the implementation a little and went for `PROPERTIES_TO_REMOVE` as you suggested.

**Comment by gmajoulet** [comment]
Thanks for the review, I changed a few things to simplify how the Stories media-pool keeps track of the video that errored. It adds its own `loadPromise` right before updating the HTMLMediaElement source, which should address the concerns previously raised.

PTAL

File: extensions/amp-story/1.0/media-tasks.js [hunk 4]
```diff
@@ -355,9 +375,10 @@ export class UpdateSourcesTask extends MediaTask {
   /**
    * @param {!Sources} newSources The sources to which the media element should
    *     be updated.
+   * @param {!Object=} options
```
Comment (jridgewell) [inline extensions/amp-story/1.0/media-tasks.js]: Make this an explicit struct.

File: src/event-helper.js [hunk 5]
```diff
@@ -156,6 +159,15 @@ export function isLoaded(eleOrWindow) {
   );
 }
 
+/**
+ * Whether the specified element/window has errored already.
+ * @param {!Element|!Window} eleOrWindow
+ * @return {boolean}
+ */
+function isErrored(eleOrWindow) {
+  return !!eleOrWindow[LOAD_FAILURE_PROPERTY];
```
Comment (jridgewell) [inline src/event-helper.js]: Can we check only the `error` attribute? That way when the media changes src, it'll clear the error out.

File: extensions/amp-story/1.0/media-tasks.js (hunk 4 above)
Comment (gmajoulet) [reply extensions/amp-story/1.0/media-tasks.js]: I've been able to get rid of it, so we no longer have to make an exception in presubmit to load `usePromise`.

File: src/event-helper.js (hunk 5 above)
Comment (gmajoulet) [reply src/event-helper.js]: The `error` attribute isn't set when the error originated from a HTMLSourceElement, and is `read-only`. :(

Comment (jridgewell) [reply src/event-helper.js]: Then let's mark the error as coming from a particular src. If the src changes, then we ignore the `LOAD_FAILURE_PROPERTY`.

Comment (gmajoulet) [reply src/event-helper.js]: Unfortunately it's not that easy, the sources could be a list of src coming from different HTMLSourceElements.

The only solution I have in mind would be to compute a string that's a list of all the `src` the browser tried before triggering the error (either the media `src`, or all its sources `src`).

If you think that's too crazy, if you're not comfortable with this change at the AMP level, I believe we could solve this reliably at the AMP Story level instead where we could easily remove that error property before changing the sources.

Comment (newmuis) [reply src/event-helper.js]: Can't amp-video do the same thing?  I feel like adding this to stories code should be a last resort here.

Comment (jridgewell) [reply src/event-helper.js]: My worry is anything else that changes the `src` (eg, amp-bind). If we don't clear the `LOAD_FAILURE_PROPERTY` there, too, then `loadPromise` will incorrectly fail for the media element.

Comment (gmajoulet) [reply src/event-helper.js]: Thanks both for the comments.
I've been able to come up with an implementation that resolves Justin's concerns, using `HTMLMediaElement.currentSrc`.
PTAL

**Comment by gmajoulet** [comment]
Friendly ping

File: src/event-helper.js [hunk 6]
```diff
@@ -170,8 +173,14 @@ export function loadPromise(eleOrWindow) {
   if (isLoaded(eleOrWindow)) {
     return Promise.resolve(eleOrWindow);
   }
+  const isMediaElement = isHTMLMediaElement(eleOrWindow);
+  if (
+    isMediaElement &&
+    eleOrWindow[MEDIA_LOAD_FAILURE_SRC_PROPERTY] === eleOrWindow.currentSrc
```
Comment (newmuis) [inline src/event-helper.js]: Can't the source be changed between when the error happens and when this executes?  Wouldn't that lead us to drop errors?

Comment (gmajoulet) [reply src/event-helper.js]: If the source changes that means the load is retried, so we want that test to fail to start listening for load/error again. I believe this is WAI, and copying the `HTMLMediaElement.error` behavior.

File: src/event-helper.js [hunk 7]
```diff
@@ -225,6 +234,14 @@ export function loadPromise(eleOrWindow) {
  *     case Windows.
  */
 function failedToLoad(eleOrWindow) {
+  // Mark the element as errored since some elements - like HTMLMediaElement
+  // using HTMLSourceElement - do not provide any synchronous way to verify if
+  // they already errored, even though the error event was already dispatched.
+  if (isHTMLMediaElement(eleOrWindow)) {
+    eleOrWindow[MEDIA_LOAD_FAILURE_SRC_PROPERTY] =
+      eleOrWindow.currentSrc || true;
```
Comment (jridgewell) [inline src/event-helper.js]: Is there a case where `currentSrc` is falsey?

Comment (gmajoulet) [reply src/event-helper.js]: When the media has no source, like `<video></video>` or `<video><source></video>`, it gives an empty string

File: extensions/amp-story/1.0/amp-story-page.js [hunk 8]
```diff
@@ -795,39 +795,47 @@ export class AmpStoryPage extends AMP.BaseElement {
...
+            .catch(unusedError => {
+              // Auto playing the media failed, which could be caused by a data
+              // saver, or a battery saving mode. Display a message so we can
+              // get a user gesture to bless the media elements, and play them.
+              if (mediaEl.tagName === 'VIDEO') {
+                this.debounceToggleLoadingSpinner_(false);
+
+                // If the error came from the <video> and not from a <source>
+                // descendant.
+                if (mediaEl.error) {
```
Comment (newmuis) [inline extensions/amp-story/1.0/amp-story-page.js]: Doesn't this also need to check `mediaEl[MEDIA_LOAD_FAILURE_SRC_PROPERTY]`?

Comment (gmajoulet) [reply extensions/amp-story/1.0/amp-story-page.js]: Yes, thanks! This test has been moved in the `loadPromise` and it'll trigger the error callback, so this (we're in the loadpromise success callback here) can be removed :)
